contracts/allocation-vester/AllocationVester.sol

```diff
@@ -0,0 +1,178 @@
+        reserve -= amount;
+
+        // remove the member from the set if its harvest has ended
+        if (member.reserve == 0) _membersAddresses.remove(msg.sender);
+
+        token.safeTransfer(msg.sender, amount);
+        emit Harvested(msg.sender, amount);
+    }
+
+    /**
+     * @notice Withdraws unallocated tokens from the contract
+     * @param amount The amount of tokens to withdraw
+     */
+    function withdraw(uint256 amount) external onlyOwner {
+        require(unreserved() >= amount, "low balance");
+        token.safeTransfer(msg.sender, amount);
+        emit Withdrawn(amount);
+    }
+
+    /**
+     * @notice Adds or removes members, or changes their allocations
+     * @param accounts An array of member addresses
+     * @param allocations The amount of tokens to allocate to the corresponding
+     * member, overriding the previous allocation
+     * @param durations The duration the corresponding allocation will last for
+     * @dev `allocation` for a member must be sufficiently greater than its `duration` such that
+     * the `rate` will be precise
+     */
+    function setAllocations(
+        address[] calldata accounts,
+        uint256[] calldata allocations,
+        uint256[] calldata durations
+    ) external onlyOwner {
+        uint256 length = accounts.length;
+        uint256 tmpReserve = reserve; // gas saving in the for loop
+        require(length != 0, "empty array");
+        require(
+            length == allocations.length && length == durations.length,
+            "varying-length arrays"
+        );
+
+        for (uint256 i; i < length; ++i) {
+            address account = accounts[i];
+            uint256 allocation = allocations[i];
+            uint256 duration = durations[i];
+            Member storage member = members[account];
+
+            require(account != address(0), "bad recipient");
+
+            uint96 tmpMemberReserve = member.reserve; // gas saving
+
+            // check the member's remaining harvest
+            if (tmpMemberReserve != 0) {
+                // stash pending rewards of the member so it remains claimable
+                uint96 tmpMemberStash = uint96(pendingHarvest(account));
+                member.stash = tmpMemberStash;
+                // free non-stashed reserves of the member from the reserves
+                tmpReserve -= (tmpMemberReserve - tmpMemberStash);
+                // free non-stashed tokens from member's reserves
+                tmpMemberReserve = tmpMemberStash;
+            }
+
+            // check the member's new allocation
+            if (allocation != 0) {
+                require(duration >= MIN_DURATION, "short vesting duration");
+
+                // lock tokens as reserve
+                tmpReserve += allocation;
+
+                // add vesting info for the member
+                tmpMemberReserve += allocation.toUint96();
+                member.rate = allocation / duration;
```

Comment:
should this be checked to verify that the rate doesn't truncate down to `0` with a small allocation or large duration?